components/feature/webchannel/src/main/java/mozilla/components/feature/webchannel/WebChannelViewFeature.kt

```diff
@@ -0,0 +1,201 @@
+import org.json.JSONArray
+import org.json.JSONObject
+import java.util.WeakHashMap
+
+/**
+ * Feature implementation that provides Firefox Accounts WebChannel support.
+ * This feature is implemented as a web extension and
+ * needs to be installed prior to use (see [WebChannelViewFeature.install]).
+ *
+ * @property context a reference to the context.
+ * @property engine a reference to the application's browser engine.
+ * @property sessionManager a reference to the application's [SessionManager].
+ */
+@Suppress("TooManyFunctions")
+class WebChannelViewFeature(
+    private val context: Context,
+    private val engine: Engine,
+    private val sessionManager: SessionManager
+) : SelectionAwareSessionObserver(sessionManager), LifecycleAwareFeature {
+
+    override fun start() {
+        observeSelected()
+
+        registerContentMessageHandler(activeSession)
+
+        if (WebChannelViewFeature.installedWebExt == null) {
+            WebChannelViewFeature.install(engine)
+        }
+    }
+
+    override fun stop() {
+        super.stop()
+    }
+
+    override fun onSessionSelected(session: Session) {
+        super.onSessionSelected(session)
+    }
+
+    override fun onSessionAdded(session: Session) {
+        registerContentMessageHandler(session)
+    }
+
+    override fun onSessionRemoved(session: Session) {
+        ports.remove(sessionManager.getEngineSession(session))
+    }
+
+    private fun sendStatusResponse(messageId: String) {
+        val statusData = JSONObject()
+        statusData.put("signedInUser", "testuser@testuser.com")
+        statusData.put("sessionToken", "ya")
+        statusData.put("uid", "uid")
+        statusData.put("verified", true)
+        statusData.put("capabilities", JSONObject().put("engines", JSONArray().put("creditcards")))
+
+        val statusMessage = JSONObject()
+        statusMessage.put("messageId", messageId)
+        statusMessage.put("command", COMMAND_STATUS)
+        statusMessage.put("data", statusData)
+
+        val status = JSONObject()
+        status.put("id", CHANNEL_ID)
+        status.put("message", statusMessage)
+
+        sendContentMessage(status)
+    }
+
+    private fun sendLinkResponse(messageId: String) {
+        val statusData = JSONObject()
+        statusData.put("ok", true)
+
+        val statusMessage = JSONObject()
+        statusMessage.put("messageId", messageId)
+        statusMessage.put("command", COMMAND_CAN_LINK_ACCOUNT)
+        statusMessage.put("data", statusData)
+
+        val status = JSONObject()
+        status.put("id", CHANNEL_ID)
+        status.put("message", statusMessage)
+
+        sendContentMessage(status)
+    }
+
+    private fun recieveLogin(messageObj: JSONObject) {
+        logger.info(messageObj.toString())
+    }
+
+    private fun registerContentMessageHandler(session: Session?) {
+        if (session == null) {
+            return
+        }
+
+        val messageHandler = object : MessageHandler {
```

Comment:
@vladikoff  You may want to take a look to @csadilek [recent PR](https://github.com/mozilla-mobile/android-components/pull/3891/files), where he fixed a memory leak on reader view  as you have a similar implementation.